src/bytecode/__init__.py

```diff
@@ -112,31 +160,59 @@ def format_instr(instr, labels=None):
                 label = labels[instr]
                 line = "%s:" % label
                 if index != 0:
-                    print()
+                    buffer.write("\n")
+            elif isinstance(instr, TryBegin):
+                line = indent + format_line(index, format_try_begin(instr, labels))
+                indent += "  "
+            elif isinstance(instr, TryEnd):
+                indent = indent[:-2]
+                line = indent + format_line(index, format_try_end(instr))
             else:
                 if instr.lineno is not None:
                     cur_lineno = instr.lineno
                 line = format_instr(instr, labels)
                 line = indent + format_line(index, line)
-            print(line)
-        print()
+            buffer.write(line + "\n")
+        buffer.write("\n")
 
     elif isinstance(bytecode, ControlFlowGraph):
         cfg_labels = {}
         for block_index, block in enumerate(bytecode, 1):
             cfg_labels[id(block)] = "block%s" % block_index
 
         for block_index, block in enumerate(bytecode, 1):
-            print("%s:" % cfg_labels[id(block)])
-            prev_lineno = None
+            buffer.write("%s:\n" % cfg_labels[id(block)])
+            seen_instr = False
             for index, instr in enumerate(block):
-                if instr.lineno is not None:
-                    cur_lineno = instr.lineno
-                line = format_instr(instr, cfg_labels)
-                line = indent + format_line(index, line)
-                print(line)
+                if isinstance(instr, TryBegin):
+                    line = indent + format_line(
+                        index, format_try_begin(instr, cfg_labels)
+                    )
+                    indent += "  "
+                elif isinstance(instr, TryEnd):
+                    if seen_instr:
+                        indent = indent[:-2]
+                    line = indent + format_line(index, format_try_end(instr))
+                else:
+                    if isinstance(instr, Instr):
+                        seen_instr = True
+                    if instr.lineno is not None:
+                        cur_lineno = instr.lineno
+                    line = format_instr(instr, cfg_labels)
+                    line = indent + format_line(index, line)
+                buffer.write(line + "\n")
             if block.next_block is not None:
-                print(indent + "-> %s" % cfg_labels[id(block.next_block)])
-            print()
+                buffer.write(indent + "-> %s\n" % cfg_labels[id(block.next_block)])
+            buffer.write("\n")
     else:
         raise TypeError("unknown bytecode class")
+
+    return buffer.getvalue()[:-1]
+
+
+def dump_bytecode(
+    bytecode: Union[Bytecode, ConcreteBytecode, ControlFlowGraph],
+    *,
+    lineno: bool = False,
+) -> None:
+    print(format_bytecode(bytecode, lineno=lineno))
```

Comment:
I added the format_bytecode that give you a str so people are free to do whatever they want with the string which I believe is enough.

Reply:
This is also not a great idea for a subtler reason: it would fail with `contextlib.redirect_stdout`, as the default is `sys.stdout`; using `None` as a default is much better.

Reply:
:) I got bitten by it when trying to package bytecode for Debian, using @fabioz's patch for https://github.com/fabioz/PyDev.Debugger where exactly this sort of construct was introduced, leaving the bytecode test suite failing. It also led to this (now-fixed) issue: https://github.com/python/cpython/issues/94286